### ISSUE TITLE
parts of POST request data is excluded

### ISSUE THREAD

**Issue by pizzaisdavid** [webrecorder/replayweb.page#61]
Hello, I am using the ~ReplayWebpage~ ArchiveWeb.page Chrome extension to capture webpages and analyze their network requests.

I have found some situations where parts of the POST data from the original request are not included in the requests of the replay.

## steps to reproduce
1. Run this script (or [clone this repo](https://github.com/pizzaisdavid/replay-webpage-post-problem), run `npm install`, and run `npm start`.)
    ```
    const express = require('express')
    const app = express()
    const port = 3000

    app.get('/test.html', (req, res) => {
      res.send(`
        <html>
          <script>
            const request = new XMLHttpRequest()
            request.open('POST', '/whatever')
            request.setRequestHeader('Content-Type', 'application/json;charset=UTF-8')
            request.send(
              JSON.stringify([{ key: 'a' }, { key: 'b' }, { key: 'c' }])
            )
          </script>
          <body>
            Page loaded. Open the network tab and refresh the page.
          </body>
        </html>
      `)
    })

    app.listen(port, () => {
      console.log(`click here: http://localhost:${port}/test.html`)
    })
    ```
1. Navigate to http://localhost:3000/test.html in a web browser.
1. Verify that the request payload is:
    ```
    [{"key":"a"},{"key":"b"},{"key":"c"}]
    ```
1. Capture the webpage with the Chrome extension (or [download the WARC from the repo](https://github.com/pizzaisdavid/replay-webpage-post-problem/blob/main/example.warc).)
1. View the WARC at https://replayweb.page
1. Verify that the request payload is:
    ```
    ...localhost:3000/whatever?__wb_method=POST&key=c
    ``` 
So instead of having `key0=a&key1=b&key2=c` (or something similar) it only takes the last value, which in this case was "c."

**Comment by pizzaisdavid** [comment]
It seems that this issue originates from the wombat package, [specifically this function](https://github.com/webrecorder/wombat/blob/7e39214b6635d70930a15a911744176caceef644/src/wombat.js#L4285), because the `q.set(k, v);` line only uses the property name and doesn't consider the entire path, which causes overriding.

I also understand that POST requests are converted to GET requests, but I don't really have the context to understand why this decision was made. To keep the conversion, I guess I would just try to fit all the properties into the URL. There is some library for encoding and decoding query strings. https://www.npmjs.com/package/query-string

**Comment by ikreymer** [comment]
Thanks for reporting! The reason the conversion happens is that it seems that, at least in Chrome, the POST data from XHR/fetch requests is otherwise dropped and not passed on to the service worker (will double check in case it has been fixed), so this was necessary to convert the POST to GET data on-the-fly.

However, you're right that there's a bug in how its being converted in wombat, not handling the duplicate params.
I actually fixed this in warcio previously, but not in wombat yet. (if you look at the WARC itself, it'll have `/whatever?__wb_method=POST&key=a&key.2_=b&key.3_=c` since it is handling it [like this](https://github.com/webrecorder/warcio.js/blob/main/src/utils.js#L131) but did not yet propagate the fix to wombat.js / replayweb.page

**Comment by pizzaisdavid** [comment]
In an attempt to help with the testing I've modified the script and got some results.
 ([also available here](https://github.com/pizzaisdavid/replay-webpage-post-problem/blob/main/index-sw.js).) But it is possible I misunderstood you, but let me know.
```
const express = require('express')
const app = express()
const port = 3000

app.get('/test.html', (req, res) => {
  res.send(`
    <html>
      <script>
        if ('serviceWorker' in navigator) {
          console.log('registering...')
          navigator.serviceWorker.register('service-worker.js');
        } else {
          console.log('not possible to register.')
        }
      </script>
      <script>
        const request = new XMLHttpRequest()
        request.open('POST', '/whatever')
        request.setRequestHeader('Content-Type', 'application/json;charset=UTF-8')
        request.send(
          JSON.stringify([{ key: 'a' }, { key: 'b' }, { key: 'c' }])
        )
      </script>
      <body>
        Yooo.. refresh the page to see console logs
      </body>
    </html>
  `)
})

app.get('/service-worker.js', (req, res) => {
  res
  .type('.js')
  .send(`
    self.addEventListener('fetch', (event) => {
      event.request.clone().json().then((body) => {
        console.log('body is...')
        console.log(body)
      })
    })
  `)
  .end()
})

app.listen(port, () => {
  console.log(`click here: http://localhost:${port}/test.html`)
})
```
![test](https://user-images.githubusercontent.com/4391884/128931545-8783e7ca-1d1d-4882-996c-79568430e22d.png)

**Comment by pizzaisdavid** [comment]
Of course you might already know this stuff, but I'll document it anyway because it took me some time to try to understand the ecosystem of Webrecorder (not saying that I understand it fully.)

Nothing would really need to change about how the webpage is initially captured with the ArchiveWebpage extension. I think this is true because I can capture a page with the latest version of the extension and load it with an [embedded ReplayWebpage thingy](https://replayweb.page/docs/embedding) using an older version before the conversion was introduced (1.3.7), and it makes a normal post request.

So when the WARC file is viewed on ReplayWebpage the frontend also loads a library called Wombat. This is responsible for intercepting the requests and directing them to the ReplayWebpage backend. I guess wombat would need to change how it handles the `POST` requests.

And then the backend would need to support the new format, which is probably Wabac. I think the conversion is handled [here](https://github.com/webrecorder/wabac.js/blob/199b90c53083415c9652f9475d453415a5088ab7/src/swmain.js#L307). Going deeper into the code, it uses the piece of code you linked earlier in Warcio, which is probably a lower level library to abstract reading and writing WARCs or something.

Don't feel obligated to correct anything here. 😄

**Comment by ikreymer** [comment]
Yes, that's partly correct, its a bit confusing :)

There were a few things, all sort of related regarding POST-to-get conversion:
- Internally, there is POST-to-get conversion happening in order to match request to response, since the matching is done via the URL
- Service workers are used for replay in wabac.js / replayweb.page The actual POST request was also being converted to get due to a bug in Chrome that dropped the post request body. That conversion had a bug that dropped duplicate arguments by using a slightly different implementation (in wombat.js)
*However*, the issue in Chrome appears to be fixed now, so disabled the conversion altogether in the latest replayweb.page for now!

- There was a slight bug in the archiveweb.page WARC export that did not serialize the POST requests in their original form, but in the converted form only. This will be fixed in the 0.6.12 release.
- The latest version of replayweb.page actually disables the POST-to-get conversion, so the original WARC you have should work as expected. However, the actual POST request in the WARC that you have is itself a 404. The 404 is now being replayed as expected. In your test app, if you fix it to handle the POST as a 200 and record that, it should replay with latest replayweb.page (1.4.7). The fix will be in the 1.4.7 release shortly as well.

Let me know if this clarifies things at all, but bottom line, is you should be able to record and replay POST requests now as expected :)

**Comment by pizzaisdavid** [comment]
Beautiful. Yep, I tested it out on the Replayweb.page site and it worked, thanks!
